rfc/20241206-oci-registries/5-providers.md

```diff
@@ -0,0 +1,100 @@
+# Provider installation in OCI
+
+---
+
+This document is part of the [OCI registries RFC](../20241206-oci-registries.md). Please leave your comments on [this pull request](https://github.com/opentofu/opentofu/pull/2163) as a review.
+
+| [« Previous](4-registry-changes.md) | [Up](../20241206-oci-registries.md) | [Next »](6-modules.md) |
+
+---
+
+As stated in [Design Considerations](3-design-considerations.md), in this iteration we will focus on serving the mirroring and private registry use case and leave solving the public OCI registry use case for a later iteration.
+
+## Configuring a provider in OCI
+
+In order to configure OCI as a provider source in OpenTofu, you will have to modify your [.tofurc](https://opentofu.org/docs/cli/config/config-file/) file. Specifically, you will need to add a `provider_installation` block with the following syntax (syntax subject to change):
+
+```hcl
+provider_installation {
+   oci {
+      registry_template = "example.com/examplenet-mirror/${namespace}-${type}"
+      include           = ["example.net/*/*"]
+   }
+   oci {
+      registry_template = "example.com/exampleorg-mirror/${namespace}-${type}"
+      include           = ["example.org/*/*"]
+   }
+   direct {
+      exclude = ["example.net/*/*", "example.org/*/*"]
+   }
+}
+```
+
+In this case, provider addresses matching `include` blocks would be redirected to the specified OCI registry. The templating here is required because OCI registry addresses work differently to OpenTofu provider addresses and some registries require specific prefixes.
+
+> [!TIP]
+> For example, Amazon ECR registries have the format of *aws_account_id*.dkr.ecr.*region*.amazonaws.com/*repository*:*tag*, which does not cleanly map to the OpenTofu provider address understanding of *hostname*/*namespace*/*type*.
+> 
+> You could work around this by configuring your organizational provider registry as follows:
+> 
+> ```hcl
+> provider_installation {
+>   oci {
+>     registry_template = "YOUR_AWS_ACCOUNT_ID.dkr.ecr.${namespace}.amazonaws.com/${type}"
+>     include           = ["yourcompany.org/*/*"]
+>   }
+> }
+> ```
+> 
+> Assuming you mirrored the AWS provider into the us-east-1 region, you could now use the following configuration:
+> 
+> ```hcl
+> terraform {
+>   required_providers {
+>     aws = {
+>       source  = "yourcompany.org/us-east-1/aws"
+>       version = "5.64.0"
+>     }
+>   }
+> }
+> ```
+
+## Storage in OCI
+
+OpenTofu takes some inspiration from how [ORAS](1-oci-primer.md#oras) stores files, but with a few key differences. It is our hope that [ORAS will soon support multi-platform images](https://github.com/oras-project/oras/issues/1053) and this implementation will be compatible.

```

Comment:
@abstractionfactory Are you looking for the multi-platform image creation support from ORAS CLI or ORAS SDK? 

For CLI, ORAS v1.3.0-beta.1 provides experimental support for multi-arch support. You can find related spec which describes the experience. https://github.com/oras-project/oras/pull/1514/

Reply:
Thanks for this pointer @FeynmanZhou!

While I think it remains true that we won't want to block on this proposal being fully implemented before we ship -- and thus I expect we'll be initially documenting a more manual process involving `oras manifest push` -- we are going to aim to be compatible with what this proposal would generate so that we can potentially update our instructions to use the proposed `oras manifest index create` (or however that finally settles) once that functionality is available in a final ORAS release.

Reply:
Thanks @abstractionfactory .